daemons/fenced/fenced_commands.c

```diff
@@ -1874,6 +1874,12 @@ can_fence_host_with_device(stonith_device_t * dev, struct device_search_s *searc
 
     } else if (pcmk__str_eq(check_type, "dynamic-list", pcmk__str_casei)) {
         time_t now = time(NULL);
+        int device_timeout = get_action_timeout(dev, "list", search->per_device_timeout);
+
+        if (device_timeout > search->per_device_timeout) {
+            crm_notice("Since the pcmk_list_timeout parameter of %s is larger than stonith-timeout(%d), timeout may occur",
+                  dev->id, search->per_device_timeout);
+        }
 
```

Comment:
Guess we don't need to do that if list comes from the cache.
Would as well reduce the notice to cases where we actually might run into the timeout as we're actually kicking off a command.
On the other hand having the action timeout here would give us the possibility to make the caching time configurable to a reasonable multiple of the the action timeout. (default action = 20s - as to end up with the same default multiple of 3 sounds reasonable)

Reply:
Hi Klaus,

> Guess we don't need to do that if list comes from the cache.
> Would as well reduce the notice to cases where we actually might run into the timeout as we're actually kicking off a command.

This means that the added check should be executed in the next if, right?

```
(snip)
    } else if (pcmk__str_eq(check_type, "dynamic-list", pcmk__str_casei)) {
        time_t now = time(NULL);

        if (dev->targets == NULL || dev->targets_age + 60 < now) {
----> Move...
            int device_timeout = get_action_timeout(dev, "list", search->per_device_timeout);

            if (device_timeout > search->per_device_timeout) {
                crm_notice("Since the pcmk_list_timeout(%ds) parameter of %s is larger than stonith-timeout(%ds), timeout may occur",
                  device_timeout, dev->id, search->per_device_timeout);
            }
(snip)
```

If so, I agree.


> On the other hand having the action timeout here would give us the possibility to make the caching time configurable to a reasonable multiple of the the action timeout. (default action = 20s - as to end up with the same default multiple of 3 sounds reasonable)

I didn't understand this opinion.
The essence of this fix is that the timeout needs to be set to stonith-timeout, pcmk_status_timeout, or pcmk_list_timeout, so no extra calculations are done.

Best Regards,
Hideo Yamauchi.

Reply:
> 
> If so, I agree.

yes, that is what I meant.

> 
> > On the other hand having the action timeout here would give us the possibility to make the caching time configurable to a reasonable multiple of the the action timeout. (default action = 20s - as to end up with the same default multiple of 3 sounds reasonable)
> 
> I didn't understand this opinion.
> The essence of this fix is that the timeout needs to be set to stonith-timeout, pcmk_status_timeout, or pcmk_list_timeout, so no extra calculations are done.

Part of what led to this fix was that pacemaker didn't respect individual timeout-settings for certain fencing-devices well enough.
Having a hardcoded 60s caching-time for dynamic-lists could be considered a similar thing.
Next best thing - regarding keeping an eye on the fencing resources needs and not necessarily on usability - to introduction of an additional caching-timeout-parameter would be using a multiple of the list-timeout.

As I was suggesting to move the code-part I was just thinking why it still could be interesting to have the list-timeout evaluated prior or as part of the decision if the list-action should be performed or not.

Regards,
Klaus

Reply:
Hi Klaus,

I would like to include this fix once.
Let's discuss the need for the addition of caching-timeout-parameters at another time.

I will implement the re-correction based on your opinion.

Many thanks,
Hideo Yamauchi.

Reply:
np - I'm fine discussing that in an extra PR - just thought it is worth thinking about and didn't want to forget

Reply:
> np - I'm fine discussing that in an extra PR - just thought it is worth thinking about and didn't want to forget

Okay!
Thanks!

Hideo Yamauchi.